src/ring_theory/polynomial/gauss_lemma.lean

```diff
@@ -43,34 +49,129 @@ begin
+  (hr : ∀ a ∈ f.roots, a ∈ T) (n : ℕ) : f.coeff n ∈ T :=
+begin
+  rw (_ : f = (f.roots.pmap (λ a h, X - C (⟨a, h⟩ : T)) hr).prod.map T.subtype),
+  { rw coeff_map, apply set_like.coe_mem },
+  conv_lhs { rw [eq_prod_roots_of_splits_id hs, hm.leading_coeff, C_1, one_mul] },
+  rw [polynomial.map_multiset_prod, multiset.map_pmap],
+  congr, convert (f.roots.pmap_eq_map _ _ hr).symm,
+  ext1, ext1, rw [polynomial.map_sub, map_X, map_C], refl,
+end
+
+variable [algebra R K]
+
+theorem frange_subset_integral_closure
+  {f : R[X]} (hf : f.monic) {g : K[X]} (hg : g.monic) (hd : g ∣ f.map (algebra_map R K)) :
+  (g.frange : set K) ⊆ (integral_closure R K).to_subring :=
+begin
+  haveI : is_scalar_tower R K g.splitting_field := splitting_field_aux.is_scalar_tower _ _ _,
+  have := coeff_mem_subring_of_splits ((splits_id_iff_splits _).2 $ splitting_field.splits g)
+    (hg.map _) (integral_closure R _).to_subring (λ a ha, roots_mem_integral_closure hf _),
+  { intros a ha, obtain ⟨n, -, rfl⟩ := mem_frange_iff.1 ha,
+    obtain ⟨p, hp, he⟩ := this n, use [p, hp],
+    rw [is_scalar_tower.algebra_map_eq R K, coeff_map, ← eval₂_map, eval₂_at_apply] at he,
+    rw eval₂_eq_eval_map, apply (injective_iff_map_eq_zero _).1 _ _ he,
+    { apply ring_hom.injective } },
+  rw [is_scalar_tower.algebra_map_eq R K _, ← map_map],
+  refine multiset.mem_of_le (roots.le_of_dvd ((hf.map _).map _).ne_zero _) ha,
+  { apply_instance },
+  { exact map_dvd (algebra_map K g.splitting_field) hd },
+  { apply splitting_field_aux.is_scalar_tower },
+end
+
+variables [is_fraction_ring R K]
+
+theorem eq_map_of_dvd [is_integrally_closed R] {f : R[X]} (hf : f.monic)
+  (g : K[X]) (hg : g.monic) (hd : g ∣ f.map (algebra_map R K)) :
+  ∃ g' : R[X], g'.map (algebra_map R K) = g :=
+begin
+  let algeq := (subalgebra.equiv_of_eq _ _ $
+    is_integrally_closed.integral_closure_eq_bot R _).trans
+    (algebra.bot_equiv_of_injective $ is_fraction_ring.injective R $ K),
+  have : (algebra_map R _).comp algeq.to_alg_hom.to_ring_hom =
+    (integral_closure R _).to_subring.subtype,
+  { ext, conv_rhs { rw ← algeq.symm_apply_apply x }, refl },
+  refine ⟨map algeq.to_alg_hom.to_ring_hom _, _⟩,
+  use g.to_subring _ (frange_subset_integral_closure hf hg hd),
+  rw [map_map, this],
+  apply g.map_to_subring,
+end
+
+lemma temporary [is_integrally_closed R] {f : R[X]} (hf : f.monic)
+  (g : K[X]) (hd : g ∣ f.map (algebra_map R K)) :
+  ∃ g' : R[X], (g'.map (algebra_map R K)) * (C $ leading_coeff g) = g :=
+begin
+  sorry
+end
 
 lemma is_primitive.is_unit_iff_is_unit_map {p : R[X]} (hp : p.is_primitive) :
   is_unit p ↔ is_unit (p.map (algebra_map R K)) :=
 hp.is_unit_iff_is_unit_map_of_injective (is_fraction_ring.injective _ _)
 
+lemma monic_eq_prod_coeff_inv {a b c : R[X]} (ha : a.monic) (h : a = b * c) :
+  leading_coeff b * leading_coeff c = 1 := sorry
+
+theorem monic.irreducible_iff_irreducible_map_fraction_map [is_integrally_closed R]
```

Comment:
This deserves a docstring (saying that it is one form of the Gauss lemma), and also to be added to the main result section at the beginning.